File: Blind 75/02 - Two Pointers/JavaScript/125-valid-palindrome.js
const { performance } = require('perf_hooks');

class Solution {
  method(s) {
    // Alphanumeric only and convert to lowercase
    s = s.replace(/[\W_]+/g, '').toLowerCase();
    let [l, r] = [0, s.length - 1];
    while (l < r) {
      if (s[l] !== s[r]) return false;
      l += 1;
      r -= 1;
    }

    return true;
  }

  reference(s) {
    if (!s.length) return true;

    s = s.toLowerCase();

    return this.isValid(s);
  }

  isValid = (s) => {
    let [left, right] = [0, s.length - 1];

    while (left < right) {
      while (left < right && this.isNonAlphaNumeric(s[left])) left++;
      while (left < right && this.isNonAlphaNumeric(s[right])) right--;

      const isSame = s[left] === s[right];
      if (!isSame) return false;

      left++;
      right--;
    }

    return true;
  };

  isNonAlphaNumeric = (char) => {
    const isNonAlpha = char < 'a' || 'z' < char; // a(97) - z(122)
    const isNonNumeric = char < '0' || '9' < char; // 0(48) - 9(57)

    return isNonAlpha && isNonNumeric;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = ['A man, a plan, a canal: Panama', 'race a car', ' '];
test.quantify(testCases);

File: Blind 75/02 - Two Pointers/125-valid-palindrome.py
'''
A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
'''
from time import time


class Solution:
    def isPalindrome(self, s: str) -> bool:
        s_formatted = ''.join(filter(str.isalnum, s)).lower()
        i = 0
        j = len(s_formatted) - 1
        while i < j:
            if s_formatted[i] != s_formatted[j]:
                return False
            i += 1
            j -= 1

        return True

    def reference(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not self.alphanum(s[l]):
                l += 1
            while l < r and not self.alphanum(s[r]):
                r -= 1
            if s[l].lower() != s[r].lower():
                return False
            l += 1
            r -= 1

        return True

    # Could write own alpha-numeric function
    def alphanum(self, c):
        return (
            ord('A') <= ord(c) <= ord('Z')
            or ord('a') <= ord(c) <= ord('z')
            or ord('0') <= ord(c) <= ord('9')
        )

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isPalindrome(case))
                else:
                    self.isPalindrome(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = ['A man, a plan, a canal: Panama', 'race a car', ' ']
    test.quantify(test_cases)
